src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSetSerializer.cs

```diff
@@ -357,11 +357,30 @@ private IEnumerable<ODataOperation> CreateODataOperations(IEnumerable<IEdmOperat
         private static Uri GetNestedNextPageLink(ODataSerializerContext writeContext, int pageSize, object obj)
         {
             Contract.Assert(writeContext.ExpandedResource != null);
+            Uri nestedNextLink = null;
             IEdmNavigationSource sourceNavigationSource = writeContext.ExpandedResource.NavigationSource;
-            NavigationSourceLinkBuilderAnnotation linkBuilder = writeContext.Model.GetNavigationSourceLinkBuilder(sourceNavigationSource);
-            Uri navigationLink =
-                linkBuilder.BuildNavigationLink(writeContext.ExpandedResource, writeContext.NavigationProperty);
-            Uri nestedNextLink = GenerateQueryFromExpandedItem(writeContext, navigationLink);
+
+            // In Contained Navigation, we don't have navigation property binding,
+            // Hence we cannot get the NavigationLink from the NavigationLinkBuilder
+            if (writeContext.NavigationSource.NavigationSourceKind() == EdmNavigationSourceKind.ContainedEntitySet)
+            {
+                // Contained navigation.
+                var linkBuilder = new NavigationSourceLinkBuilderAnnotation(sourceNavigationSource, writeContext.Model);
+                var idlink = linkBuilder.BuildIdLink(writeContext.ExpandedResource);
+
+                var link = idlink.ToString() + "/" + writeContext.NavigationSource.Name;
+                nestedNextLink = new Uri(link);
```

Comment:
Should this logic be pushed down to where the link is actually created, rather than you having to fix it up here?  i.e., should this fix really be made as a special case for NavigationSourceKind.ContainedEntitySet in LinkGenerationHelpers.GenerateBaseODataPathSegments?  It seems likely that there are other methods that may trying to generate links that are broken for contained navigation properties?